Various macOS UI improvements

vit9696 opened the pull request:
Hello,

This pull-request consolidates and fixes most of the existing changes (e.g. #4757 and #2096) and brings us other improvements. Whereas this is not a complete fix to all design issues in qBittorrent in regards to macOS, this is certainly a good start.

Please let me know what you consider important to change (if anything), and let's get this merged :)

The changelist includes:

- [x] Removed icons from main and right-click menus (thx #4757);
- [x] Removed icons from buttons (this also is invalid in macOS);
- [x] Removed tray statusbar support from macOS (thx #4757);
- [x] Removed Hide and Quit duplicate entries from Dock menu (thx #4757);
- [x] Reduced horizontal category indentation (it looks terrible on macOS and does nothing);
- [x] Enabled Dock minimisation instead of quitting on window close (thx #4757, fixed window restoration);
- [x] Replaced toolbar separators by spacers, there is no such concept on macOS;
- [x] Added download speed badge to dock icon (thx #2096);
- [x] Fixed awkward tab widget border and spacing looking terrible;
- [x] Fixed horizontal toolbar appearing in preferences list;
- [x] Renamed Options to Preferences in the toolbar and window title (thx #3018);
- [x] Removed ambiguous window icons;
- [x] Fixed multiple torrent addition windows being inaccessible (thx #6206).

<details>
  <summary>Screenshots</summary>

![before](https://user-images.githubusercontent.com/4348897/27018853-464ec676-4f3c-11e7-8b1b-bb1ffff46b15.png)
![after](https://user-images.githubusercontent.com/4348897/27018855-49ec7436-4f3c-11e7-9c73-f05ff0370272.png)

</details>

Regarding separator → spacer replacement: I am thinking of removing them entirely, since they do not really add much clarity, but to my eyes it looks fine either way. If you think it is better to remove them, let me know.

Regarding Dock restoration: it is a very annoying bug (or shall I say feature) in Qt, no event is delivered to the dispatcher when you click on the icon regardless of the docs. The only reason is to use native APIs, and although it is a little ugly, it works well.

Regarding the badges: the API Qt offers is very limited, and one cannot display more than a single line, and its length is very short. To avoid dotted shortening even the "D: " prefix was removed. Unfortunately it also does not allow colouring, and to avoid different sorts of confusion I decided not to add the uploading speed badge.

Even though I am open for a discussion on what is worth changing and may even make another pull-request fixing certain other areas, I will not add any new features to the current one.

Among the macOS-specific issues I am aware of are all the menu/app icons. They are supposed to be greyscale, not coloured. And I am actually wondering how would you prefer to change it?

vit9696 commented:
I fixed the build issues except the one with qmake+gui. Your homebrew formula defines `-skip qtmacextras`, which is necessary to provide Dock icon badges and thus mandatory. Once you update your toolchain (just removing this line should be enough), your buildsystem will start working once again.

I could make it build even without the badges, but I see no reason to limit macOS functionality just because a certain toolchain is broken.

evsh commented on src/CMakeLists.txt:
```diff
@@ -14,12 +14,23 @@ find_package(LibtorrentRasterbar REQUIRED)
 find_package(Qt5 5.5.1 COMPONENTS ${QBT_QT_COMPONENTS} REQUIRED)
 
+if (GUI AND APPLE)
+    include_directories(${Qt5MacExtras_INCLUDE_DIRS})
```
This one and the next 2 lines shall not be needed, as `target_link_libraries(... Qt5::MacExtras)`, initiated by `list (APPEND QBT_QT_COMPONENTS MacExtras)` call above should do the very same.

vit9696 replied:
I thought this myself, but actually it does not work. If I remove these lines, compilation will stop working.
I could not explain the reason and simply hardcoded the variables. I think it is a bug in Qt, because I saw other people doing similar things on the net.

evsh replied:
You are right, I forgot that when we migrated from Qt4, I replaced smart commands with direct component names. You have to mention Qt5::MacExtras in target_link_libraries() directly. For instance:
```cmake
if (APPLE)
	target_link_libraries(qbt_gui Qt5::MacExtras)
endif()
```
in src/gui/CMakeLists.txt

vit9696 replied:
Indeed, thanks.

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -1929,7 +1929,7 @@ QIcon MainWindow::getSystrayIcon() const
     }
 
     QIcon icon;
-#if (defined(Q_OS_UNIX)
```
why this fails the compilation?

Chocobo1 commented on src/app/main.cpp:
```diff
@@ -252,6 +252,11 @@ int main(int argc, char *argv[])
 
+#ifdef Q_OS_MAC
+        // On OS X the standard is to not show icons in the menus
+        app->setAttribute(Qt::AA_DontShowIconsInMenus);
```
there are already an `#if defined(Q_OS_MAC)` block at line 210
you should move it there.

Chocobo1 commented on src/base/preferences.h:
```diff
@@ -115,16 +115,18 @@ class Preferences: public QObject
     void setHideZeroValues(bool b);
     int getHideZeroComboValues() const;
     void setHideZeroComboValues(int n);
+#ifndef Q_OS_MAC
```
I would take out all entries in this `#ifndef` block and move it down to where you put
```c++
#ifndef Q_OS_MAC
     TrayIcon::Style trayIconStyle() const;
```

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -52,6 +52,11 @@
 #include <QCryptographicHash>
 #include <QProcess>
 
+#ifdef Q_OS_MAC
```
move this block under
```
#include "mainwindow.h"
```

Chocobo1 replied:
and move
```c++
#include <QtMac>
#include <QtMacExtras>
```
to here

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -204,6 +211,9 @@ MainWindow::MainWindow(QWidget *parent)
     qDebug("create tabWidget");
     m_tabs = new HidableTabWidget(this);
     connect(m_tabs, SIGNAL(currentChanged(int)), this, SLOT(tabChanged(int)));
+#ifdef Q_OS_MAC
```
move this block 1 line up.

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -251,6 +261,18 @@ MainWindow::MainWindow(QWidget *parent)
+    
+#ifdef Q_OS_MAC
+    foreach (QAction *action, m_ui->toolBar->actions()) {
+        if (action->isSeparator()) {
```
what its purpose?

vit9696 replied:
This removes vertical separators from the toolbar menu, they are invalid in macOS design. As I said above, there is no such concept. One could have space separation or no separation.

I replaced the separators with some extra space, but I think they are fine even if just removed.

Chocobo1 replied:
OK, after searching for some toolbar screenshot on google, I get what you mean.

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -343,6 +367,7 @@ MainWindow::MainWindow(QWidget *parent)
     // Load Window state and sizes
     readSettings();
 
+#ifndef Q_OS_MAC
```
is this equivalent? 
```c++
    if (!m_systrayIcon) {
        // Make sure the Window is visible if we don't have a tray icon
        if (pref->startMinimized()) {
            showMinimized();
        }
        else {
            show();
            activateWindow();
            raise();
        }
    }
#ifndef Q_OS_MAC
    else {
        if (!(pref->startMinimized() || m_uiLocked)) {
            show();
            activateWindow();
            raise();
        }
        else if (pref->startMinimized()) {
            showMinimized();
            if (pref->minimizeToTray())
                hide();
        }
    }
#endif
```

vit9696 replied:
m_systrayIcon is undefined on mac, so one cannot write it this way.

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -1022,12 +1055,20 @@ void MainWindow::showEvent(QShowEvent *e)
 void MainWindow::closeEvent(QCloseEvent *e)
 {
     Preferences *const pref = Preferences::instance();
+#ifndef Q_OS_MAC
```
change it to `#ifdef Q_OS_MAC` and switch the code

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -1395,9 +1489,15 @@ void MainWindow::updateGUI()
     }
+#else
+    if (status.payloadDownloadRate > 0)
+        QtMac::setBadgeLabelText(tr("%1/s").arg(Utils::Misc::friendlyUnit(status.payloadDownloadRate)));
```
change `tr` to `QString`

vit9696 replied:
Are you sure this is correct suggestion? E.g. in Russian I would want to see %1/с not s. Since seconds in Russian are `секунды`.

Chocobo1 replied:
>Are you sure this is a correct suggestion?

After reading your explanation, obviously not.
IMO, it would be better to add this:
```c++
QtMac::setBadgeLabelText(tr("%1/s", "s is shorthand for seconds")
    .arg(Utils::Misc::friendlyUnit(status.payloadDownloadRate)));
```
and split the line please.

vit9696 replied:
Alright, done

Chocobo1 commented on src/gui/mainwindow.h:
```diff
@@ -102,23 +105,24 @@ class MainWindow: public QMainWindow
     void balloonClicked();
     void writeSettings();
     void readSettings();
+#ifndef Q_OS_MAC
```
move this block down to before `private slots:`

Chocobo1 commented on src/gui/optionsdlg.cpp:
```diff
@@ -143,13 +143,17 @@ OptionsDialog::OptionsDialog(QWidget *parent)
 
     // Load options
     loadOptions();
+#ifndef Q_OS_MAC
```
write `#ifdef Q_OS_MAC` switch code.

Chocobo1 commented:
2 things I wanted to add:
1. create fixup commits for now, note that eventually you'll have to squash all commits into one.
2. please provide before & after screenshots if possible, majority (at least me) of the devs don't have easy access to a mac.

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -1191,6 +1234,50 @@ void MainWindow::dragEnterEvent(QDragEnterEvent *event)
+    Class cls = objc_getClass("NSApplication");
+    objc_object *appInst = objc_msgSend(reinterpret_cast<objc_object *>(cls), sel_registerName("sharedApplication"));
+    
+    if (appInst != nullptr) {
```
```c++
if (!appInst) 
  return;
```

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -1191,6 +1234,50 @@ void MainWindow::dragEnterEvent(QDragEnterEvent *event)
+
+static MainWindow *dockMainWindowHandle;
+
+static bool dockClickHandler(id self, SEL _cmd, ...) {
```
move `{` to a newline.

Chocobo1 replied:
pls avoid using underscore in variable names.

Chocobo1 replied:
>pls avoid using underscore in variable names.

@vit9696 
I think you missed this one.

vit9696 replied:
Yep, actually did. Fixed it now.

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -1191,6 +1234,50 @@ void MainWindow::dragEnterEvent(QDragEnterEvent *event)
+    return false;
+}
+
+void MainWindow::setupDockClickHandler() {
```
move { to a newline.

thalieht commented on src/gui/categoryfilterwidget.cpp:
```diff
@@ -73,6 +73,7 @@ CategoryFilterWidget::CategoryFilterWidget(QWidget *parent)
     setIconSize(Utils::Misc::smallIconSize());
 #if defined(Q_OS_MAC)
     setAttribute(Qt::WA_MacShowFocusRect, false);
+    setIndentation(0);
```
I don't know if it's any different in mac but in KDE if you do this and enable `Use Subcategories` in options, any subcategories vanish from the sidepanel and there doesn't seem to be a way to show them because the expand icon isn't visible.

I realize this is probably out of the scope of this PR, IMO this should be used on all platforms but only when `Use Subcategories` is disabled and maybe reduce the indentation of branches (and the root as much as possible).

vit9696 replied:
@thalieht actually I faield to find this option yesterday, and thus assumed that subcategories simply do not work on macOS. Now I found it, thanks :)

I changed the indentation to not be disabled when they are off, but doing something like this for all the platforms is the job of a separate pull request indeed.

vit9696 commented:
@Chocobo1 I fixed all the issues you asked me to change except the ones I consider invalid.
The screenshots are available in the first post if you did not notice ^^

thalieht commented on src/gui/categoryfilterwidget.cpp:
```diff
@@ -73,6 +73,8 @@ CategoryFilterWidget::CategoryFilterWidget(QWidget *parent)
     setIconSize(Utils::Misc::smallIconSize());
 #if defined(Q_OS_MAC)
     setAttribute(Qt::WA_MacShowFocusRect, false);
+    if (!BitTorrent::Session::instance()->isSubcategoriesEnabled())
```
But if `Use Subcategories` is changed it would need a restart for the indentation to change. Is that ok?
If you can make it change without restart i would like to suggest to move this out of the `#if` so all platforms can have it.

vit9696 replied:
Of course I could fix it, and I have just done it. Thanks for reminding. But you are right, if I change it outside the preprocessor macro, it wouldn't fit this PR.

I have no development tools installed for the other platforms, but I think it should work fine for them. You or anybody else could submit a new PR removing the ifdefs once this gets merged :)

Chocobo1 commented on src/gui/categoryfilterwidget.h:
```diff
@@ -57,4 +57,8 @@ private slots:
     QString askCategoryName();
+
+#ifdef Q_OS_MAC
+    int defaultIndentation;
```
private variables should prepended with `m_`,
thus `m_defaultIndentation`

vit9696 replied:
Fixed

vit9696 commented:
Alright, I squashed all the changes into a single commit and even fixed a couple of other issues I spotted. Would you please merge this?

Chocobo1 commented on src/app/main.cpp:
```diff
@@ -208,14 +208,17 @@ int main(int argc, char *argv[])
+            QByteArray path = "/usr/local/bin:";
+            path += qgetenv("PATH");
+            qputenv("PATH", path.constData());
+        }
```
please do us a favor if you don't mind: remove the curly braces here, it isn't really needed.

vit9696 replied:
ok

Chocobo1 commented:
>Would you please merge this?

I would like to if I can verify this PR on a mac, otherwise, this PR will be handled by the project owner sledgehammer999.

vit9696 commented:
@Chocobo1 well, I build this on a vm (too much of a dislike for Qt to let it arrive on my real machines), and can give you vnc or teamviewer without any issues. Maybe you could spot something yourself…

LordNyriox commented:
@vit9696:  After this PR is merged, might you work on updating [<https://github.com/qbittorrent/qBittorrent/pull/3018>]?  It includes some things (like renaming the `Options` GUI to `Preferences` on Mac) which might go well with the consistency tweaks in this PR.  

Thank you for your awesome work!

vit9696 commented:
@LordNyriox actually it is a very small change, so I added it right here.
It is especially bad because of the main menus, where the entry already says "Preferences…", and because of Dock that has Options.

To be honest, after this I felt like making all the icons greyscale on macOS and be done with it. I am not sure how to do it properly. On macOS one pretty much must use greyscale icons on toolbar, statusbar, in fact everywhere. And what you have at the moment looks terrible.

Shall I open an issue for a discussion? You have some theming pull requests here awaiting the merge, but this is not really about theming but about following the mandatory UI guidelines.

vit9696 commented:
And just another thing, while I was changing the Preferences title I thought that something is wrong with the window. Now I see it. You use qBittorrent icon everywhere in the windows.

On macOS we **do not** have a window logo unless this logo represents a file type, which is quite convenient. Also fixed.

vit9696 commented:
In addition to the screens in the first post, here are more:
<details>
  <summary>Screenshots</summary>

![screen shot 2017-06-13 at 02 30 40](https://user-images.githubusercontent.com/4348897/27076183-16d3eaee-5035-11e7-8d69-a896905f89a0.png)
![screen shot 2017-06-13 at 02 30 48](https://user-images.githubusercontent.com/4348897/27076187-1727ccd6-5035-11e7-9685-3576b302e7e9.png)
![screen shot 2017-06-13 at 02 31 03](https://user-images.githubusercontent.com/4348897/27076184-172730c8-5035-11e7-9316-699516eda2e7.png)
![screen shot 2017-06-13 at 02 32 18](https://user-images.githubusercontent.com/4348897/27076188-172bf0b8-5035-11e7-8043-55820d909c97.png)
![screen shot 2017-06-13 at 02 32 28](https://user-images.githubusercontent.com/4348897/27076185-172729b6-5035-11e7-8a6b-118cf9235275.png)
![screen shot 2017-06-13 at 02 33 43](https://user-images.githubusercontent.com/4348897/27076186-1727cbbe-5035-11e7-91e3-32d0f2eb78b7.png)

</details>

LordNyriox commented:
@vit9696:  Regarding making the icons greyscale:  

1.  For me, personally, that was one of the silliest changes in Lion.  Mac OS X was so colorful that I could identify icons from color alone—then Lion comes out with its UI simplification and suddenly the only difference between sidebar icons is shape?  Ugh.  

2.  As for how to do it, two approaches are possible: generate a greyscale version of each icon used in Mac, and use an `#ifdef` to use those on Mac OS—or paint the icons grey in-process, using an `#ifdef` to define when to do so.  

3.  If you do pursue using greyscale icons on Mac, please make it optional (perhaps reusing the Linux `Use system icons` option or something similar).  

4.  Actually, most Mac users I know would suggest you leave the icons colored.  The greyscale change was relatively recent, after all, and not all apps adhere to it (including some first-party apps, in fact).  Mind you, I have not tested Sierra (only El Capitan), so I cannot speak to any UI changes it has—but I doubt I am the only one using an older and more colorful version of the OS.  

5.  As for discussing this on a separate thread, go ahead!  That would really be a better place to discuss such a drastic change than here.  

Thanks again!

LordNyriox commented on src/gui/mainwindow.cpp:
```diff
@@ -262,7 +309,9 @@ MainWindow::MainWindow(QWidget *parent)
     connect(m_ui->actionDecreasePriority, SIGNAL(triggered()), m_transferListWidget, SLOT(decreasePrioSelectedTorrents()));
     connect(m_ui->actionBottomPriority, SIGNAL(triggered()), m_transferListWidget, SLOT(bottomPrioSelectedTorrents()));
+#ifndef Q_OS_MAC
     connect(m_ui->actionToggleVisibility, SIGNAL(triggered()), this, SLOT(toggleVisibility()));
```
@vit9696:  Not directly relevant to this PR, but do you think you could add a commit updating the `connect` syntax used in `mainwindow.cpp` to the Qt5 standard?  

There should be a dedicated function for that in Qt Creator, if I recall correctly.

vit9696 replied:
Sorry, but I am not very familiar with Qt, and I do not even have Qt Creator installed.

Quick googling shows me this: https://wiki.qt.io/New_Signal_Slot_Syntax, and by looking at the syntax I am afraid all your code requires such changes, and not just the mac section.

Therefore it is pretty much outside of the scope of this pull request and is better to be done by a Qt dev.

vit9696 commented:
I created an issue regarding macOS icon choice: https://github.com/qbittorrent/qBittorrent/issues/6957.
Please let me hear your opinion.

@LordNyriox I am a reasonably old macOS user, and from what I remember even in the days of 10.6 colouring was not something to appear in the toolbar, and there were hardly many programs with the statusbar, and even them had greyscale icons. (e.g. [2017-06-13 16 09 25](https://user-images.githubusercontent.com/4348897/27087288-459a8d10-505d-11e7-8315-2d91617db7ba.png)). I guess we could continue the discussion in the created issue.

Chocobo1 commented on src/gui/optionsdlg.cpp:
```diff
@@ -95,7 +99,13 @@ OptionsDialog::OptionsDialog(QWidget *parent)
-        m_ui->tabSelection->item(i)->setSizeHint(QSize(std::numeric_limits<int>::max(), 64));  // uniform size for all icons
+        // uniform size for all icons
+#ifdef Q_OS_MAC
+        // Smaller size for mac due to rendering specifics to avoid vertical scroll
```
you can either do this change for all platform (no need or `#ifdef`)
or change the options dialog default height (enlarge the number `<height>591</height>` in optionsdlg.ui)

vit9696 replied:
Hm, but will this be fine? I do not see many reasons to increase the total height, and I am not sure if reducing the size is ok on all the platforms.

Chocobo1 replied:
both way will be fine, increasing the dialog height is a safer way to go (if you only need adding little more height).

vit9696 replied:
To be honest, I do not think that increasing the dialogue height is a good idea. It already is 615 pixels tall here, plus 75 pixels for the Dock and 25 for the menubar. This makes it pretty big already for smaller screens (e.g. when remoting from iPad, which has 1024x768 res). I reduced the icon size instead.

vit9696 commented:
Sorry, found a bug, which was previously unnoticed due to another bug in the very old code of yours, that started to trigger due to a Qt change @_@.

The issue is that recent qBittorrent binaries have black text in selected tabs (see [Search and Suzumiya](https://user-images.githubusercontent.com/4348897/27076188-172bf0b8-5035-11e7-8043-55820d909c97.png). This happens because to support fusion and other styles on mac (that have black text on white tabs) Qt devs added a stylesheet comparison in https://github.com/qt/qtbase/commit/2fbc1432025dd478cf9e940e3c2a1952b218c6dc.

In your codebase you have this [stylesheet replacement](https://github.com/qbittorrent/qBittorrent/commit/7b57345c4f9307fc11d6ea7e61ea38802df92ed4) that… does nothing on mac mhm. (It was later modified and moved a little but remained the same). This effectively switches all the styles on mac to Stylesheet. QMacStyle check fails, and starting with the recent Qt 5.8.x versions qBittorrent tabs on mac are unreadable.

After I fixed this I discovered that I myself added a stylesheet change to remove the border. This is not valid, it breaks all the tabwidgets and `-style fusion` mode as well. I reworked the change, so we are good now ([screen 1](https://user-images.githubusercontent.com/4348897/27253348-6c93dfea-537b-11e7-9f88-d0fb1f21eddf.png) and [screen 2](https://user-images.githubusercontent.com/4348897/27253349-6cdea4f8-537b-11e7-807c-2a66b31226f6.png)).

I squashed the commits, just as you asked, but here are the only changes I made since the review confirmation to make your life easier: https://ghostbin.com/paste/umfdu.

Chocobo1 commented:
>I squashed the commits, just as you asked, but here are the only changes I made since the review confirmation to make your life easier: https://ghostbin.com/paste/umfdu.

Yeah, if you could defer the squash-ups to just before merging, that would be great, example: https://github.com/qbittorrent/qBittorrent/pull/6911/commits

Chocobo1 commented on src/gui/hidabletabwidget.h:
```diff
@@ -43,6 +43,16 @@ class HidableTabWidget : public QTabWidget
 
 protected:
+#ifdef Q_OS_MAC
+    void paintEvent(QPaintEvent* event)
```
`void paintEvent(QPaintEvent* event) override`

vit9696 replied:
Oh, you actually do use override and this was not intentional… Right, fixed for the rest as well.

Chocobo1 commented on src/gui/hidabletabwidget.h:
```diff
@@ -43,6 +43,16 @@ class HidableTabWidget : public QTabWidget
+    void paintEvent(QPaintEvent* event)
+    {
+        // Hide the pane for macintosh style
+        if (!style()->inherits("QMacStyle")) {
```
I would remove the curly brackets

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -205,6 +218,11 @@ MainWindow::MainWindow(QWidget *parent)
 
+#ifdef Q_OS_MAC
+    // Increase top spacing to avoid tab overlapping
+    m_ui->centralWidgetLayout->addSpacing(8);
```
there is `m_ui->centralWidgetLayout->addWidget(m_tabs);` below, is it possible to move this block to there?

vit9696 replied:
Maybe it is indeed more readable.

vit9696 commented:
As for squashing I am fine not to, generally even the maintainer does it at merging (on this site it is even automated). However, if I have to squash "in the end" it will most likely decline the confirmed review, so it is kind of an infinite loop with this logic. 

I left the commits untouched, let me know if we are good so that I could squash them and this gets finally applied xD))

Chocobo1 commented:
>However, if I have to squash "in the end" it will most likely decline the confirmed review, so it is kind of an infinite loop with this logic.

Don't worry about it, its how we run the project now...

>I left the commits untouched, let me know if we are good so that I could squash them and this gets finally applied xD))

Someone (from the devs) will give you the signal.

---

@sledgehammer999 
ping and I my review is done here.

thalieht commented:
Closes #951, #2126, half #2761, #3030, #3617, #3750?, #4129?, #4200, #4748?, PR #2096

vit9696 commented:
Thanks for checking the tracker.

This indeed fixes #4748.

As for #2761 using just the bare API there is enough space for a single line only, I guess for a more complete fix one must perform a full manual redraw in Cocoa, or at least provide an option to choose the speed to show. Given the enforced red colour the latter is probably a bad idea, and the former is certainly outside of this PR.

#3750, #4129, #4147 are fixed, the completely broken tray icon is now gone for good, and restoration and minimisation works fine.

sledgehammer999 commented:
@vit9696 thanks for awesome work and sorry for the long wait. At least the other members started reviewing the code.
Of course I approve this work visually.
I'll also update the toolchain on Travis as soon as I get my hands on the macOS machine. Usually in a few days, where I'll also build the upcoming v3.3.14.

PS: With my latest push your PR has a minor conflict. Please fix it. Thx.

vit9696 commented:
Hi, nice to see that moving. I rebased everything to trunk.
Had to add a minor change to get cmake compile: [line](https://github.com/qbittorrent/qBittorrent/pull/6952/files#diff-b565ef9558c99a298c3a692809dba35cL29).
Hopefully this could get merged in 3.4.0 =)

evsh commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -214,7 +214,11 @@ void AddNewTorrentDialog::show(QString source, const BitTorrent::AddTorrentParam
             ok = dlg->loadTorrent(source);
 
         if (ok)
+#ifdef Q_OS_MAC
```
I would like to see the non-modal dialog on other systems as well.

vit9696 replied:
Good idea, but definitely out of the scope of this PR, especially because I do not have the dev environment set for mac. I could imagine someone removing the ifdefs for both non-modal dialogues and subcategory padding =)

vit9696 commented on src/CMakeLists.txt:
```diff
@@ -17,12 +17,21 @@ if (GUI)
 find_package(Qt5 5.5.1 COMPONENTS ${QBT_QT_COMPONENTS} REQUIRED)
 
+if (GUI AND APPLE)
+    list (APPEND QBT_GUI_OPTIONAL_LINK_LIBRARIES ${Qt5MacExtras_LIBRARIES})
```
Are those related https://github.com/qbittorrent/qBittorrent/pull/6952#pullrequestreview-43354865?

evsh replied:
Yes. `target_link_libraries(qbt_gui Qt5::MacExtras)` provides "-I", "-L", and "-l" compiler switches.

vit9696 replied:
Checked it, builds fine without this line. Removed. Thanks =)

vit9696 commented:
> I don't like all these #ifdef with icons. Maybe we better organize our icons to remove or blacklist action icons on MacOS?

Even though possibly a good idea, it is not really doable. Icons still remain in the search tab, for example. While not so nice, this is essential for usability.

evsh commented:
> Even though possibly a good idea, it is not really doable.

Could you elaborate, please? If you mean that the same icon might be used on other platforms in different places where on MacOS in one of them icon is not needed, then maybe we better add a context parameter to `GuiIconProvider::getIcon()`?

vit9696 commented:
@evsh I was referring to blacklisting tab/button icons as a class on macOS. I think Qt has something similar for non-menu icons that disables them globally, which is not really possible, since in some places the icons are still needed even on macOS.

Wrapping functions like `GuiIconProvider::getIcon()` is possible, but from what I remember it is not the only function used, also the semantics of making a show/hide decision needs to be agreed on (it cannot just be ifdef macOS → hide). In addition there is other code that still needs ifdefs.

All in all, I would rather not try to refactor relevant side portions of code at the same time as fixing a lot of other things.

sledgehammer999 commented:
@vit9696 rebase on master. macOS packageds for travis/homebrew have been updated to include QtMacExtras (+ QtSVG) (plus latest RC_1_0 for libtorrent)

vit9696 commented:
@sledgehammer999 we are good now =)

evsh commented:
> macOS packageds for travis/homebrew have been updated to include QtMacExtras (+ QtSVG)

Why do we use a custom Qt bottle but not the package from homebrew?

LordNyriox commented:
@evsh:  

>Why do we use a custom Qt bottle but not the package from homebrew? 

I seem to recall a discussion some time ago that resulted in this decision.  

Something about packaging errors or something like that, IIRC.

vit9696 commented:
Now that everything is settled including the build issues, would you please approve and merge this or let me know what else is ought to be changed? It is not too hard to maintain the patch in sync with the ongoing changes in master, but I would rather not spend your and my time on stuff like rebasing. Thank you for understanding.

CC @Chocobo1, @evsh, @sledgehammer999

evsh commented on src/CMakeLists.txt:
```diff
@@ -17,12 +17,20 @@ if (GUI)
 
+if (GUI AND APPLE)
+    # Fix MOC inability to detect macOS
+    add_definitions(-DQ_OS_MAC)
```
Why this is needed and why in GUI build only?

vit9696 replied:
This is to workaround Qt bugs, examples: [one](https://bugreports.qt.io/browse/QTBUG-52687), [two](https://bugreports.qt.io/browse/QTBUG-58325). From what I can tell there are no proper solutions.

It is necessary for [this file](https://github.com/qbittorrent/qBittorrent/pull/6952/files#diff-cd08514626207f11a4741106092c8404R37) in particular, which is GUI only. Without this define MOC tries to generate systray functions, and even if I let it include QSystemTrayIcon it will later fail with linking errors.

evsh replied:
But both bugs are marked as resolved. Why do you care about Qt < 5.9.0?

evsh replied:
I mean Homebrew installs the latest Qt. How a user can get the old version?

vit9696 replied:
This bugs are marked as resolved, but they are not resolved as of Qt 5.9.1 under our circumstances.

evsh replied:
I can not find the same define for qmake build. How does it build then? Also, if this a fix for a specific file and bug, would it be possible to pass the flag for that file target only supplying a comment why the flag is needed?

vit9696 replied:
I suppose Qt team fixed the bug for qmake actually (which could be the reason they closed the issue), or your custm Qt build has patches from HomeBrew. It certainly fails with cmake though.

The flag is specific for the header, so there are many cpp files. It is saner to define it for all the files, since this flag is defined everywhere but in MOC for all macOS sources, and not defining it may lead to undefined behaviour in other places.

If you think a different comment is more descriptive, would you suggest anything? Maybe adding some Qt bug links?

evsh replied:
> It is saner to define it for all the files, since this flag is defined everywhere but in MOC for all macOS sources, and not defining it may lead to undefined behaviour in other places.

Nevertheless, I would be as specific as possible. Why don't you use [CMAKE_AUTOMOC_MOC_OPTIONS](https://cmake.org/cmake/help/v3.9/variable/CMAKE_AUTOMOC_MOC_OPTIONS.html#variable:CMAKE_AUTOMOC_MOC_OPTIONS) variable?

> If you think a different comment is more descriptive, would you suggest anything? Maybe adding some Qt bug links?

Exactly. Please, add such a comment.

vit9696 replied:
Well, I was simply unaware of this feature. Thanks for pointing it out, added.

sledgehammer999 commented:
>Why do we use a custom Qt bottle but not the package from homebrew?

In the past the `macdeployqt` was broken. But now homebrew includes the patch in the formular itself. I can't remember any other reason. So it is probably safe to drop it now.

>Now that everything is settled including the build issues, would you please approve and merge this or let me know what else is ought to be changed? It is not too hard to maintain the patch in sync with the ongoing changes in master, but I would rather not spend your and my time on stuff like rebasing. Thank you for understanding.

I leave this PR open in a browser tab. I will review+merge as soon as possible.

evsh commented:
> In the past the macdeployqt was broken. But now homebrew includes the patch in the formular itself. I can't remember any other reason. So it is probably safe to drop it now.

Great! Thanks!

vit9696 commented:
> In the past the macdeployqt was broken. But now homebrew includes the patch in the formular itself. I can't remember any other reason. So it is probably safe to drop it now.

It may not be. I cannot produce a standalone build with `macdeployqt` from HomeBrew, though I am unaware if I am supposed to pass special flags or compile differently. Please test it before removing =).

sledgehammer999 commented:
>It may not be. I cannot produce a standalone build with macdeployqt from HomeBrew, though I am unaware if I am supposed to pass special flags or compile differently. Please test it before removing =).

Are you using the 5.9.1 `qt.rb` formula? It explicitly has patch for this inside. And used `otool -L` in the QtNetwork and QtGui libs and the QtCore references were rewritten to relative paths pointing to the .app package.

PS: I just use `macdeployqt qbittorrent.app`

vit9696 commented:
@sledgehammer999 I have just rechecked my setup, and there is good news. Qt from HomeBrew now deploys without issues. There still were 5.9.0 leftovers, and unfortunately the older macdeployqt was used and confused me.

There is an important thing to be aware of, though: Qt from HomeBrew requires 10.10 and higher.

sledgehammer999 commented:
>There is an important thing to be aware of, though: Qt from HomeBrew requires 10.10 and higher.

We don't actually have a policy for macOS. Are there people not upgraded to 10.10 yet?

vit9696 commented:
@sledgehammer999 unsure, there could be some who stick to 10.9.5, but I personally do not care. Just mentioned this because you were concerned about the minimum supported macOS version in the other issue.

sledgehammer999 commented on src/gui/mainwindow.cpp:
```diff
@@ -1086,6 +1159,7 @@ void MainWindow::createTorrentTriggered(const QString &path)
 bool MainWindow::event(QEvent *e)
 {
     switch (e->type()) {
+#ifndef Q_OS_MAC
```
Why don't you ifdef the whole switch here?

sledgehammer999 commented on src/gui/optionsdlg.cpp:
```diff
@@ -95,7 +99,8 @@ OptionsDialog::OptionsDialog(QWidget *parent)
     for (int i = 0; i < m_ui->tabSelection->count(); ++i) {
-        m_ui->tabSelection->item(i)->setSizeHint(QSize(std::numeric_limits<int>::max(), 64));  // uniform size for all icons
+        // uniform size for all icons
+        m_ui->tabSelection->item(i)->setSizeHint(QSize(std::numeric_limits<int>::max(), 62));
```
Why did you change it to 62?

vit9696 replied:
To let it fit the window without showing a vertical scrollbar. See https://github.com/qbittorrent/qBittorrent/pull/6952/files/8f48837ed9fe4799c9f917ff22eab6226371d03d#diff-2c91a19e5c2af0562178cfdaacafa3ec.
I fixed the above one as well, was probably due to an automatic merge of some PR.

sledgehammer999 replied:
OK.
The link doesn't work for me.
Merge the new commit into the **1st** one. And also rebase.
Then I can merge it.

vit9696 replied:
Done =)

sledgehammer999 commented:
Thx for the time you took to find all the little details and fix them.

Chocobo1 commented:
Getting some errors on qbt launch:
```shell
$ src/qbittorrent 
QObject::connect: No such slot MainWindow::toggleVisibility()
QObject::connect:  (sender name:   'actionToggleVisibility')
QObject::connect:  (receiver name: 'MainWindow')
QObject::connect: No such slot MainWindow::updateTrayIconMenu()
QObject::connect:  (receiver name: 'MainWindow')
QObject::connect: No such slot MainWindow::toggleVisibility(QSystemTrayIcon::ActivationReason)
QObject::connect:  (receiver name: 'MainWindow')
```

vit9696 commented:
@Chocobo1 I definitely don't have this. The only issues I see are:
```
$ /Applications/qbittorrent.app/Contents/MacOS/qbittorrent ; exit; 
2017-07-16 12:46:45.118 qbittorrent[34665:31468133] Path  given to -[NSWorkspace iconForFile:] is not a full path.
2017-07-16 12:46:45.118 qbittorrent[34665:31468133] Path  given to -[NSWorkspace iconForFile:] is not a full path.
```
And they are most certainly due to https://github.com/qbittorrent/qBittorrent/pull/6156.
Are you using cmake or qmake builds? I use cmake, and now I do wonder whether it also needs Q_OS_MAC definition for MOC. Like [this one](https://github.com/qbittorrent/qBittorrent/pull/6952/files#diff-95e351a3805a1dafa85bf20b81d086e6R30).

Chocobo1 commented:
@vit9696 
No worries, #7106 has just been merged.

>Are you using cmake or qmake builds?  

qmake

vit9696 commented:
Oh, I see, thanks a lot :)

evsh commented on src/CMakeLists.txt:
```diff
@@ -17,12 +17,21 @@ if (GUI)
+if (GUI AND APPLE)
+    # Fix MOC inability to detect macOS. This seems to only affect cmake.
+    # Relevant issue: https://bugreports.qt.io/browse/QTBUG-58325
+    set(CMAKE_AUTOMOC_MOC_OPTIONS ${CMAKE_AUTOMOC_MOC_OPTIONS} -DQ_OS_MAC)
```
@vit9696, could you test CMake 3.9 and without this define, please? As of version 3.9 CMake generates `moc_predefs.h` as qmake does.

vit9696 replied:
Tested, exactly the same error. Find reveals no such file `moc_predefs.h`.

evsh replied:
Thank you. The results is quite unexpected.

D3V4N5H commented:
Hello. I'm relatively new. I love qBittorrent. I'd like to see ETA (expected time to finish download) on the Badge instead of Global Download Speed, as it makes sense to me, to not to check now and then from the window and then hide it again.

I know C++ but not Qt.
How can I contribute?
Do I need any special development environment in my MacBook Air?
